core/drivers/ls_i2c.c

```diff
@@ -0,0 +1,404 @@
+	{ 2560, 0xB0 },	 { 3072, 0xB1 },  { 3584, 0xB2 }, { 3840, 0xAF },
+	{ 4096, 0xB3 },	 { 4608, 0xB4 },  { 5120, 0xB8 }, { 6144, 0xB9 },
+	{ 7168, 0xBA },	 { 7680, 0xB7 },  { 8192, 0xBB }, { 9216, 0xBC },
+	{ 10240, 0xBD }, { 12288, 0xBE }, { 15360, 0xBF }
+};
+
+/*
+ * I2C divisor and ibfd register values when glitch filter is disabled.
+ * In case of duplicate SCL divisor value, the ibfd value with high MUL value
+ * has been selected. A higher MUL value results in a lower sampling rate of
+ * the I2C signals. This gives the I2C module greater immunity against glitches
+ * in the I2C signals.
+ */
+static const struct i2c_clock_divisor_pair
+	clk_div_glitch_disabled[] = {
+	{ 20, 0x0 },	 { 22, 0x1 },	  { 24, 0x2 },	  { 26, 0x3 },
+	{ 28, 0x8 },	 { 30, 0x5 },	  { 32, 0x9 },	  { 34, 0x6 },
+	{ 36, 0x0A },	 { 40, 0x40 },	  { 44, 0x41 },	  { 48, 0x42 },
+	{ 52, 0x43 },	 { 56, 0x48 },	  { 60, 0x45 },	  { 64, 0x49 },
+	{ 68, 0x46 },	 { 72, 0x4A },	  { 80, 0x80 },	  { 88, 0x81 },
+	{ 96, 0x82 },	 { 104, 0x83 },	  { 112, 0x88 },  { 120, 0x85 },
+	{ 128, 0x89 },	 { 136, 0x86 },	  { 144, 0x8A },  { 160, 0x8B },
+	{ 176, 0x8C },	 { 192, 0x90 },	  { 208, 0x56 },  { 224, 0x91 },
+	{ 240, 0x1F },	 { 256, 0x92 },	  { 272, 0x8F },  { 288, 0x93 },
+	{ 320, 0x98 },	 { 352, 0x95 },	  { 384, 0x99 },  { 416, 0x96 },
+	{ 448, 0x9A },	 { 480, 0x5F },	  { 512, 0x9B },  { 576, 0x9C },
+	{ 640, 0xA0 },	 { 768, 0xA1 },	  { 896, 0xA2 },  { 960, 0x9F },
+	{ 1024, 0xA3 },	 { 1152, 0xA4 },  { 1280, 0xA8 }, { 1536, 0xA9 },
+	{ 1792, 0xAA },	 { 1920, 0xA7 },  { 2048, 0xAB }, { 2304, 0xAC },
+	{ 2560, 0xAD },	 { 3072, 0xB1 },  { 3584, 0xB2 }, { 3840, 0xAF },
+	{ 4096, 0xB3 },	 { 4608, 0xB4 },  { 5120, 0xB8 }, { 6144, 0xB9 },
+	{ 7168, 0xBA },	 { 7680, 0xB7 },  { 8192, 0xBB }, { 9216, 0xBC },
+	{ 10240, 0xBD }, { 12288, 0xBE }, { 15360, 0xBF }
+};
+
+/*
+ * software reset of the entire I2C module.
+ * The module is reset and disabled.
+ * Status register fields (IBSR) are cleared.
+
+ * @param[in] Base       Base Address of I2c controller's registers
+ */
+void i2c_reset(vaddr_t base)
+{
+	struct i2c_regs *regs = (struct i2c_regs *)base;
+
+	io_setbits8((vaddr_t)&regs->ibcr, I2C_IBCR_MDIS);
+	io_setbits8((vaddr_t)&regs->ibsr, I2C_IBSR_IBAL | I2C_IBSR_IBIF);
+	io_clrbits8((vaddr_t)&regs->ibcr, I2C_IBCR_IBIE | I2C_IBCR_DMAEN);
+	io_clrbits8((vaddr_t)&regs->ibic, I2C_IBIC_BIIE);
+}
+
+static uint8_t i2c_get_ibfd(vaddr_t base, uint16_t clock_divisor)
+{
+	struct i2c_regs *regs = (struct i2c_regs *)base;
+	uint8_t ibfd = 0; /* I2c Bus Frequency Divider Register */
+	const struct i2c_clock_divisor_pair *dpair = NULL;
+	size_t dpair_sz = 0;
+	unsigned int n = 0;
+
+	if (io_read8((vaddr_t)&regs->ibdbg) & I2C_IBDBG_GLFLT_EN) {
+		dpair = clk_div_glitch_enabled;
+		dpair_sz = ARRAY_SIZE(clk_div_glitch_enabled);
+	} else {
+		dpair = clk_div_glitch_disabled;
+		dpair_sz = ARRAY_SIZE(clk_div_glitch_disabled);
+	}
+
+	if (clock_divisor > dpair[dpair_sz - 1].divisor) {
```

Comment:
Line 118-129 could be simplified as:
``` C
        for (n = 0; n < dpair_sz - 1; n++)
                if (clock_divisor < dpair[n].divisor)
                        break; 
        return dpair[n].ibfd;
```
I believe it would be easier to get the intention of the code too. Don't you agree?

Reply:
Yes, I agree with you. Thanks for pointing out.